overlord/snapstate/snapstate.go

```diff
@@ -2296,7 +2297,42 @@ func canRemove(st *state.State, si *snap.Info, snapst *SnapState, removeAll bool
 		rev = si.Revision
 	}
 
-	return PolicyFor(si.Type(), deviceCtx.Model()).CanRemove(st, snapst, rev, deviceCtx)
+	if err := PolicyFor(si.Type(), deviceCtx.Model()).CanRemove(st, snapst, rev, deviceCtx); err != nil {
+		return err
+	}
+
+	// check if this snap is required by any validation set in enforcing mode
+	enforcedSets, err := EnforcedValidationSets(st)
+	if err != nil {
+		return err
+	}
+	if enforcedSets != nil {
+		requiredValsets, requiredRevision, err := enforcedSets.CheckPresenceRequired(si)
+		if err != nil {
+			if _, ok := err.(*snapasserts.PresenceConstraintError); !ok {
```

Comment:
I checked the implementation of `CheckPresenceRequired()` but I did not understand in which case the snap constraint can be `asserts.PresenceInvalid`. If it actually means that the snap presence is invalid (that is, that the snap should not be allowed to be present), then I think we should return `nil` here, otherwise we could be stuck in an invalid situation.